### DISCUSSION TITLE
Should I be seeing a list of available actions when calling `embark-act`?

### DISCUSSION THREAD

**Discussion by apc** [emacs-citar/citar#88] [Q&A] [answered]
Ivy user here, testing the waters. Thanks for setting this up. Had shied away from trying non-ivy packages because `ivy-bibtex` is a central part of my workflow. 

I've now disabled all my `ivy`-related packages and setup `selectrum`, `embark`, `consult`, and `bibtex-actions`. If I try e.g. `bibtex-actions-open` I get a list of entries in my main bibliography file. But calling `embark-act` on any of them doesn't show me a list of options, which is what I would have expected. I see that typing e.g. `o` will open the URL associated with a given entry, but looking at the `bibtex-actions-keymap` doesn't help me figure out how to e.g. edit the associated note etc. 

Sorry if I'm missing something obvious here! 

**EDIT**: OK, I figured out where to get the values of `bibtex-actions-keymap` in a readable form. (I just needed to go to the source!). For reasons that escape me, calling `C-h v` on `bibtex-actions-keymap` shows me the following unhelpful list of values: 

```
bibtex-actions-map is a variable defined in ‘bibtex-actions.el’.
Its value is shown below.

  This variable may be risky if used as a file-local variable.

Documentation:
Keymap for ’bibtex-actions’.

Value:
(keymap
 (97 . bibtex-actions-add-pdf-to-library)
 (101 . bibtex-actions-open-entry)
 (110 . bibtex-actions-open-notes)
 (116 . bibtex-actions-add-pdf-attachment)
 (98 . bibtex-actions-insert-bibtex)
 (107 . bibtex-actions-insert-key)
 (114 . bibtex-actions-insert-reference)
 (99 . bibtex-actions-insert-citation)
 (108 . bibtex-actions-open-link)
 (112 . bibtex-actions-open-pdf)
 (111 . bibtex-actions-open))

[back]

```

I was unaware that looking at `bibtex-actions.el` in this case would have shown a human readable value. 

I still want to know whether I *should* be seeing a list of available actions when calling `embark-act`, and if not whether there's a way of having that show up. I realize this may be something having to do with my `embark` settings; if so, please just let me know and I'll close this. (I just copied the example settings from the README on the `embark` repository.)

**Comment by bdarcus** [comment, marked as answer]
Hi @apc - answer to your topic question is YES!

It definitely should work as you expect. The first screenshot on the README is what you should see if you have `which-key` setup and after invoking `embark-act`.

First, questions:

Did you set the embark variable mentioned in the README:

https://github.com/bdarcus/bibtex-actions#basic

It occurs to me it's possible that section is too brief ATM.

If, as I assume, yes, then two additional questions:

1. do you have marginalia installed and active? I think this is necessary for embark to be able to associate these commands with that keymap.
2. do you have embark configured to load the menus? You may have to do something there.

Once setup, it should just work.

Edit: I just added these test scripts, which are explained in the README.

https://github.com/bdarcus/bibtex-actions/tree/main/test

You can look through them for some config ideas.

**Comment by bdarcus** [comment]
PS: you don't actually need to deactivate ivy in your config. You just need to turn it off, and turn on selectrum, etc.

**Comment by apc** [comment]
Hi @bdarcus, thanks for your quick reply! This is my current setup for `bibtex-actions`: 

```
(use-package bibtex-actions
  :demand t
  :config
  (setf (alist-get 'bibtex embark-keymap-alist) 'bibtex-actions-map)
  (setq bibtex-completion-bibliography '("~/bib/apcmain.bib")))
```

I figured out that if I type `C-h` after calling `embark-act` I see the list of available bindings. Before doing that, all I see is this: 

<img width="1283" alt="image" src="https://user-images.githubusercontent.com/350504/114720664-3507cb80-9d06-11eb-9348-b0f2734c7ab7.png">

This may well be the intended behavior. Coming from `ivy-bibtex`, though, I was used to seeing the menu of options when calling `ivy-dispatching-done` from the list of `bib` entries. Maybe that's why I got confused? 

I'll think about your question re: the README section. In the meantime, here are a few thoughts. 

- I had previously set all variables from `bibtex-completion` (including `bibtex-completion-bibliography`) in my `use-package` declaration for `bibtex`. I suspect most people trying out `bibtex-actions` may have set those values in their init. I could have used a reminder to move all those settings to the `use-package` declaration for `bibtex-actions`. So, something like: "You may have already set values for many of the `bibtex-completion` variables that `bibtex-actions` relies on; you want to make sure those variables are set the way you want before you first call any of the `bibtex-actions-` commands."
- A sample `use-package` declaration, listing all the values that are needed for most of the actions to work as expected (`bibtex-completion-library-path`, `bibtex-completion-bibliography`, etc.) would be nice to have. 
- It might be helpful to explain that you need to have `embark` loaded for the setting of `embark-keymap-alist` not to give an error. (So, perhaps having `:after embark` in the example declaration?)
- Finally, just a line saying "to see the list of available `bibtex-actions`, call `embark-bindings`" would have been super helpful. (Or perhaps something much like [this section](https://github.com/tmalsburg/helm-bibtex#actions-for-selected-publications) of the `helm-bibtex` documentation?)

**Comment by apc** [comment]
PS: when I call `embark-bindings` I see a list of available actions for the selected entry, but it's embedded in this much longer listing of actions available. Is there some way of just seeing `bibtex-actions` related bindings?

**Comment by bdarcus** [comment]
> PS: when I call `embark-bindings` I see a list of available actions for the selected entry, but it's embedded in this much longer listing of actions available. Is there some way of just seeing `bibtex-actions` related bindings?

You shouldn't need to call `embark-binding` if this is setup right.

Does something like this setup, from the selectrum test scripts, work for you?

https://github.com/raxod502/selectrum/blob/master/test/selectrum-orderless.el

**Comment by bdarcus** [comment]
> I'll think about your question re: the README section. In the meantime, here are a few thoughts.

Thank you!

What do you think about explicitly including how to get this running using the selectrum test scripts, and then moving the more detailed examples to the wiki?

**Comment by apc** [comment]
> > I'll think about your question re: the README section. In the meantime, here are a few thoughts.
> 
> Thank you!
> 
> What do you think about explicitly including how to get this running using the selectrum test scripts, and then moving the more detailed examples to the wiki?

That would be great!

**Comment by apc** [comment]
> > PS: when I call `embark-bindings` I see a list of available actions for the selected entry, but it's embedded in this much longer listing of actions available. Is there some way of just seeing `bibtex-actions` related bindings?
> 
> You shouldn't need to call `embark-binding` if this is setup right.
> 
> Does something like this setup, from the selectrum test scripts, work for you?
> 
> https://github.com/raxod502/selectrum/blob/master/test/selectrum-orderless.el

That helps! I ended up adding

```
(setq embark-prompter #'embark-completing-read-prompter)
(global-set-key (kbd "M-o") #'embark-act)
```

and that gave me a much more user-friendly experience. I'd still like to be able to only see the list of `bibtex-actions` available, rather than all the options I now see. Do you know if there's an easy way to make that happen?

**Comment by bdarcus** [comment]
> I'd still like to be able to only see the list of `bibtex-actions` available, rather than all the options I now see. Do you know if there's an easy way to make that happen?

I strongly agree!

Two things:

First, I had a recent chat with embark author @oantolin about just this, and he noted it's on his todo.

https://github.com/oantolin/embark/issues/197#issuecomment-815300932

Would it help for one of us to add an explicit issue @oantolin?

Second, in the meantime, I sort the commands in the menu, so the `bibtex-actions` ones come first.

https://github.com/bdarcus/bibtex-actions/wiki/Configuration#which-key

That's for `which-key` though; not sure if or how to do that with the default one.

**Comment by bdarcus** [comment]
BTW, marginalia is also generally useful, beyond its connection to embark.

**Comment by apc** [comment]
> > I'd still like to be able to only see the list of `bibtex-actions` available, rather than all the options I now see. Do you know if there's an easy way to make that happen?
> 
> I strongly agree!
> 
> Two things:
> 
> First, I had a recent chat with embark author @oantolin about just this, and he noted it's on his todo.
> 
> [oantolin/embark#197 (comment)](https://github.com/oantolin/embark/issues/197#issuecomment-815300932)
> 
> Would it help for one of us to add an explicit issue @oantolin?
> 
> Second, in the meantime, I sort the commands in the menu, so the `bibtex-actions` ones come first.
> 
> https://github.com/bdarcus/bibtex-actions/wiki/Configuration#which-key
> 
> That's for `which-key` though; not sure if or how to do that with the default one.

Very helpful! I should have checked the wiki first. Using `which-key` is fine by me for now. (I take it the `which-key` settings work because `bibtex-` appears first when actions are sorted alphabetically, right?)

And yes, I've installed `marginalia` as well. I'm still very new to all of this, though, so I'm sure there's plenty for me to learn before I can get things where I wanted.

**Comment by bdarcus** [comment]
> I take it the `which-key` settings work because `bibtex-` appears first when actions are sorted alphabetically, right?)

Right. I can't say I chose the package name for that reason though!

> And yes, I've installed `marginalia` as well. I'm still very new to all of this, though, so I'm sure there's plenty for me to learn before I can get things where I wanted.

There's a moderate learning curve to move from monolithic solutions like ivy and helm to deliberately decomposed, but collaboratively developed, solutions like these.

It probably took me two weeks to "get it".

Many of the decisions here reflect now having understood it. My hope is the wiki can flourish as users start to experiment.

**Comment by bdarcus** [comment]
Last question: would it help others, do you think, to move this to the discussion forum?

It's easy enough for me to do.

**Comment by apc** [comment]
Yes, I think so. In general, I find it much easier to learn to use packages by first simply copying from an opinionated setup and then tweaking it to my taste rather than building from scratch. I suspect I'm not alone in this!

**Reply by bdarcus** [reply]
It's now on the discussion forum.

I'll work on the README soon.

**Reply by bdarcus** [reply]
PS - I use Doom, which partly explains why I don't have more default `use-package` details ATM.

But I link to my doom config in #86.
